File: public/js/blocks.js
export const BlockId = {
  AIR: 0,
  GRASS: 1,
  DIRT: 2,
  STONE: 3,
  WOOD: 4,
  LEAVES: 5,
  SAND: 6,
  WATER: 7,
  BEDROCK: 8,
  COBBLESTONE: 9,
  PLANKS: 10,
  GLASS: 11,
  SNOW: 12,
  CRAFTING_TABLE: 13,
  LAVA: 14,
  OBSIDIAN: 15,
  TALL_GRASS: 16,
  FLOWER: 17,
  WHEAT: 18,
};

export const BLOCKS = {
  [BlockId.AIR]: {
    id: BlockId.AIR,
    name: 'Air',
    solid: false,
    transparent: true,
    color: null,
  },
  [BlockId.GRASS]: {
    id: BlockId.GRASS,
    name: 'Grass',
    solid: true,
    transparent: false,
    color: { top: 0x5a9e3a, side: 0x8b6914, bottom: 0x8b6914 },
  },
  [BlockId.DIRT]: {
    id: BlockId.DIRT,
    name: 'Dirt',
    solid: true,
    transparent: false,
    color: 0x8b6914,
  },
  [BlockId.STONE]: {
    id: BlockId.STONE,
    name: 'Stone',
    solid: true,
    transparent: false,
    color: 0x888888,
  },
  [BlockId.WOOD]: {
    id: BlockId.WOOD,
    name: 'Wood',
    solid: true,
    transparent: false,
    color: { top: 0xc4a35a, side: 0x6b4423, bottom: 0xc4a35a },
  },
  [BlockId.LEAVES]: {
    id: BlockId.LEAVES,
    name: 'Leaves',
    solid: true,
    transparent: true,
    color: 0x2d6b1e,
  },
  [BlockId.SAND]: {
    id: BlockId.SAND,
    name: 'Sand',
    solid: true,
    transparent: false,
    color: 0xe8d5a3,
  },
  [BlockId.WATER]: {
    id: BlockId.WATER,
    name: 'Water',
    solid: false,
    transparent: true,
    color: 0x3366cc,
    opacity: 0.6,
  },
  [BlockId.BEDROCK]: {
    id: BlockId.BEDROCK,
    name: 'Bedrock',
    solid: true,
    transparent: false,
    color: 0x333333,
  },
  [BlockId.COBBLESTONE]: {
    id: BlockId.COBBLESTONE,
    name: 'Cobblestone',
    solid: true,
    transparent: false,
    color: 0x6b6b6b,
  },
  [BlockId.PLANKS]: {
    id: BlockId.PLANKS,
    name: 'Planks',
    solid: true,
    transparent: false,
    color: 0xc4a35a,
  },
  [BlockId.GLASS]: {
    id: BlockId.GLASS,
    name: 'Glass',
    solid: true,
    transparent: true,
    color: 0xc8e8ff,
    opacity: 0.3,
  },
  [BlockId.SNOW]: {
    id: BlockId.SNOW,
    name: 'Snow',
    solid: true,
    transparent: false,
    color: 0xffffff,
  },
  [BlockId.CRAFTING_TABLE]: {
    id: BlockId.CRAFTING_TABLE,
    name: 'Crafting Table',
    solid: true,
    transparent: false,
    color: { top: 0xc4a35a, side: 0x8b6914, bottom: 0x6b4423 },
  },
  [BlockId.LAVA]: {
    id: BlockId.LAVA,
    name: 'Lava',
    solid: false,
    transparent: true,
    color: 0xff5500,
    emissive: true,
    damages: true,
  },
  [BlockId.OBSIDIAN]: {
    id: BlockId.OBSIDIAN,
    name: 'Obsidian',
    solid: true,
    transparent: false,
    color: 0x1a0a2e,
  },
  [BlockId.TALL_GRASS]: {
    id: BlockId.TALL_GRASS,
    name: 'Tall Grass',
    solid: false,
    transparent: true,
    color: 0x4a9e32,
  },
  [BlockId.FLOWER]: {
    id: BlockId.FLOWER,
    name: 'Flower',
    solid: false,
    transparent: true,
    color: { top: 0xffdd55, side: 0x3d8a2f, bottom: 0x3d8a2f },
  },
  [BlockId.WHEAT]: {
    id: BlockId.WHEAT,
    name: 'Wheat',
    solid: false,
    transparent: true,
    color: { top: 0xd4b84a, side: 0xc4a035, bottom: 0x8b6914 },
  },
};

export const HOTBAR_BLOCKS = [
  BlockId.GRASS,
  BlockId.DIRT,
  BlockId.STONE,
  BlockId.COBBLESTONE,
  BlockId.WOOD,
  BlockId.PLANKS,
  BlockId.LEAVES,
  BlockId.OBSIDIAN,
  BlockId.GLASS,
];

export function isSolid(blockId) {
  return BLOCKS[blockId]?.solid ?? false;
}

export function isTransparent(blockId) {
  return BLOCKS[blockId]?.transparent ?? true;
}

export function isLava(blockId) {
  return blockId === BlockId.LAVA;
}

export function getBlockColor(blockId, face) {
  const block = BLOCKS[blockId];
  if (!block?.color) return 0xffffff;
  if (typeof block.color === 'number') return block.color;
  if (face === 'top') return block.color.top;
  if (face === 'bottom') return block.color.bottom;
  return block.color.side;
}

File: rust/src/blocks.rs
use bevy::prelude::*;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
#[repr(u8)]
pub enum BlockId {
    #[default]
    Air = 0,
    Grass = 1,
    Dirt = 2,
    Stone = 3,
    Wood = 4,
    Leaves = 5,
    Sand = 6,
    Water = 7,
    Bedrock = 8,
    Cobblestone = 9,
    Planks = 10,
    Glass = 11,
    Snow = 12,
    CraftingTable = 13,
    Lava = 14,
    Obsidian = 15,
    TallGrass = 16,
    Flower = 17,
    Wheat = 18,
}

impl BlockId {
    pub fn from_u8(v: u8) -> Self {
        match v {
            1 => Self::Grass,
            2 => Self::Dirt,
            3 => Self::Stone,
            4 => Self::Wood,
            5 => Self::Leaves,
            6 => Self::Sand,
            7 => Self::Water,
            8 => Self::Bedrock,
            9 => Self::Cobblestone,
            10 => Self::Planks,
            11 => Self::Glass,
            12 => Self::Snow,
            13 => Self::CraftingTable,
            14 => Self::Lava,
            15 => Self::Obsidian,
            16 => Self::TallGrass,
            17 => Self::Flower,
            18 => Self::Wheat,
            _ => Self::Air,
        }
    }

    pub fn is_cross_decoration(self) -> bool {
        matches!(self, Self::TallGrass | Self::Flower | Self::Wheat)
    }

    /// Blocks that impede player movement (full cubes only).
    pub fn blocks_collision(self) -> bool {
        match self {
            Self::TallGrass | Self::Flower | Self::Wheat | Self::Leaves | Self::Glass => false,
            _ => self.solid(),
        }
    }

    pub fn solid(self) -> bool {
        matches!(
            self,
            Self::Grass
                | Self::Dirt
                | Self::Stone
                | Self::Wood
                | Self::Leaves
                | Self::Sand
                | Self::Bedrock
                | Self::Cobblestone
                | Self::Planks
                | Self::Glass
                | Self::Snow
                | Self::CraftingTable
                | Self::Obsidian
        )
    }

    pub fn transparent(self) -> bool {
        !self.solid() || matches!(self, Self::Leaves | Self::Glass | Self::Water | Self::Lava)
    }

    pub fn color(self, face: Face) -> Color {
        let rgb = match self {
            Self::Air => return Color::NONE,
            Self::Grass => match face {
                Face::Top => [0.35, 0.62, 0.24],
                Face::Bottom => [0.55, 0.41, 0.08],
                Face::Side => [0.55, 0.41, 0.08],
            },
            Self::Dirt => [0.55, 0.41, 0.08],
            Self::Stone => [0.53, 0.53, 0.53],
            Self::Wood => match face {
                Face::Top | Face::Bottom => [0.77, 0.64, 0.35],
                Face::Side => [0.42, 0.27, 0.14],
            },
            Self::Leaves => [0.18, 0.42, 0.12],
            Self::Sand => [0.91, 0.84, 0.64],
            Self::Water => [0.20, 0.40, 0.80],
            Self::Bedrock => [0.20, 0.20, 0.20],
            Self::Cobblestone => [0.42, 0.42, 0.42],
            Self::Planks => [0.77, 0.64, 0.35],
            Self::Glass => [0.78, 0.91, 1.0],
            Self::Snow => [1.0, 1.0, 1.0],
            Self::CraftingTable => match face {
                Face::Top => [0.77, 0.64, 0.35],
                Face::Bottom => [0.42, 0.27, 0.14],
                Face::Side => [0.55, 0.41, 0.08],
            },
            Self::Lava => [1.0, 0.33, 0.0],
            Self::Obsidian => [0.10, 0.04, 0.18],
            Self::TallGrass => [0.29, 0.62, 0.20],
            Self::Flower => match face {
                Face::Top => [1.0, 0.87, 0.33],
                Face::Bottom | Face::Side => [0.24, 0.54, 0.18],
            },
            Self::Wheat => match face {
                Face::Top => [0.83, 0.72, 0.29],
                Face::Bottom | Face::Side => [0.77, 0.63, 0.21],
            },
        };
        let shade = match face {
            Face::Top => 1.0,
            Face::Bottom => 0.6,
            Face::Side => 0.8,
        };
        let emissive = self == Self::Lava;
        let scale = if emissive { 1.5 } else { 1.0 } * shade;
        Color::srgb(rgb[0] * scale, rgb[1] * scale, rgb[2] * scale)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Face {
    Top,
    Bottom,
    Side,
}

pub fn block_drop(id: BlockId) -> Option<BlockId> {
    match id {
        BlockId::Grass => Some(BlockId::Dirt),
        BlockId::Stone => Some(BlockId::Cobblestone),
        BlockId::Leaves | BlockId::TallGrass | BlockId::Flower | BlockId::Wheat => None,
        BlockId::Air | BlockId::Water | BlockId::Lava | BlockId::Bedrock => None,
        other => Some(other),
    }
}

pub fn is_block_item(id: u16) -> bool {
    (1..=99).contains(&id)
}
